### PR TITLE
fix(arns): disallow negative token costs

### PR DESCRIPTION
**Pull request by dtfiedler**

### CODE REVIEW

File: src/utils.lua [hunk 1]
```diff
@@ -7,17 +7,6 @@ function utils.hasMatchingTag(tag, value)
...
--- Then, check for a 43-character base64url pattern.
--- The pattern checks for a string of length 43 containing alphanumeric characters, hyphens, or underscores.
-function utils.isValidBase64Url(url)
-	local isValidBase64Url = #url == 43 and string.match(url, "^[%w-_]+$") ~= nil
-
-	if not isValidBase64Url then
-		error("String pattern is invalid.")
-	end
-	return url
-end
```
Comment (dtfiedler) [inline src/utils.lua:19]: cruft from aos copy/pasta - we use more specific utils and this has been confused with those

File: spec/arns_spec.lua [hunk 2]
```diff
@@ -516,23 +516,25 @@ describe("arns", function()
 				purchaseType = "lease",
 				years = 2,
 				name = "test-name",
+				currentTimestamp = timestamp,
 			}
 			_G.DemandFactor.currentDemandFactor = demandFactor
 			assert.are.equal(expectedCost, arns.getTokenCost(intendedAction))
 		end)
-		it("should return the correct token cost for a permabuy", function()
+		it("should return the correct token cost for a buying permanently owned", function()
```
Comment (arielmelendez) [inline spec/arns_spec.lua:524]: funny grammar

File: spec/arns_spec.lua [hunk 3]
```diff
@@ -574,10 +576,68 @@ describe("arns", function()
 				intent = "Extend-Lease",
 				years = 2,
 				name = "test-name",
+				currentTimestamp = timestamp + constants.oneYearMs,
 			}
 			_G.DemandFactor.currentDemandFactor = demandFactor
 			assert.are.equal(expectedCost, arns.getTokenCost(intendedAction))
 		end)
+		it("should throw an error if trying to increase undername limit an a record in the grace period", function()
```
Comment (arielmelendez) [inline spec/arns_spec.lua:584]: typo: `and`

File: spec/arns_spec.lua [hunk 4]
```diff
@@ -574,10 +576,68 @@ describe("arns", function()
...
+				intent = "Increase-Undername-Limit",
+				quantity = 5,
+				name = "test-name",
+				currentTimestamp = 10000000 + 1, -- in the grace period
+			})
+			assert.is_false(status)
+			assert.match("Name must be active to increase undername limit", error)
+		end)
+		it(
+			"should throw an error if trying to increase undername limit on a record that is expired beyond its grace period",
```
Comment (arielmelendez) [inline spec/arns_spec.lua:603]: nit: weird whitespacing

File: spec/arns_spec.lua [hunk 5]
```diff
@@ -802,7 +862,7 @@ describe("arns", function()
 		end)
 
 		describe("getPriceForAuctionAtTimestamp", function()
-			it("should return the correct price for an auction at a given timestamp for a permabuy", function()
+			it("should return the correct price for an auction at a given timestamp for permanently owned", function()
```
Comment (arielmelendez) [inline spec/arns_spec.lua:865]: The find and replace all makes the grammar on some of these kinda funny.

File: spec/arns_spec.lua (hunk 4 above)
Comment (dtfiedler) [reply spec/arns_spec.lua:603]: linting rule for line length

File: src/arns.lua [hunk 6]
```diff
@@ -361,73 +370,59 @@ function arns.getTokenCost(intendedAction)
...
 	assert(type(intent) == "string", "Intent is required and must be a string.")
 	assert(type(name) == "string", "Name is required and must be a string.")
 	if intent == "Buy-Record" then
-		assert(purchaseType == "lease" or purchaseType == "permabuy", "PurchaseType is invalid.")
-		if purchaseType == "lease" then
-			assert(years >= 1 and years <= 5, "Years is invalid. Must be an integer between 1 and 5")
-		end
+		-- stub the process id as it is not required for this intent
+		local processId = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
+		arns.assertValidBuyRecord(name, years, purchaseType, processId)
```
Comment (fedellen) [inline src/arns.lua:381]: this part seems odd to be in the non-test code. without digging into the function I wonder why wouldn't `assertValidBuyRecord` take an optional processID instead

Comment (fedellen) [reply src/arns.lua:381]: I guess you're just getting the cost here, whereas the assertion function may be used in later on in the creation flow

File: spec/arns_spec.lua (hunk 5 above)
Comment (dtfiedler) [reply spec/arns_spec.lua]: addressed in [da8ab39](https://github.com/ar-io/ar-io-network-process/pull/156/commits/da8ab39db7266c47d5652789cbbb17c4af73464b) and [a2bf822](https://github.com/ar-io/ar-io-network-process/pull/156/commits/a2bf8227d1ca5341b12c161c34eaf7ab35628ab2)

File: spec/arns_spec.lua (hunk 3 above)
Comment (dtfiedler) [reply spec/arns_spec.lua]: addressed in [da8ab39](https://github.com/ar-io/ar-io-network-process/pull/156/commits/da8ab39db7266c47d5652789cbbb17c4af73464b) and [a2bf822](https://github.com/ar-io/ar-io-network-process/pull/156/commits/a2bf8227d1ca5341b12c161c34eaf7ab35628ab2)

File: spec/arns_spec.lua (hunk 2 above)
Comment (dtfiedler) [reply spec/arns_spec.lua]: addressed in [da8ab39](https://github.com/ar-io/ar-io-network-process/pull/156/commits/da8ab39db7266c47d5652789cbbb17c4af73464b) and [a2bf822](https://github.com/ar-io/ar-io-network-process/pull/156/commits/a2bf8227d1ca5341b12c161c34eaf7ab35628ab2)

**Review by fedellen — APPROVED** [review]

File: src/arns.lua [hunk 7]
```diff
@@ -361,73 +370,59 @@ function arns.getTokenCost(intendedAction)
...
-
-		if record.type == "permabuy" then
-			error("Name is already a permabuy")
-		end
+		arns.assertValidUpgradeName(record, currentTimestamp)
 		tokenCost = arns.calculatePermabuyFee(baseFee, demand.getDemandFactor())
 	end
+	-- if token Cost is less than 0, throw an error
+	if tokenCost < 0 then
+		error("Invalid token cost for " .. intendedAction.intent)
```
Comment (arielmelendez) [inline src/arns.lua:400]: Might be helpful for debugging if the bad value were in the error string too.

File: src/main.lua [hunk 8]
```diff
@@ -1813,7 +1815,7 @@ addEventingHandler(ActionMap.ReassignName, utils.hasMatchingTag("Action", Action
 	local initiator = utils.formatAddress(msg.Tags.Initiator)
 	local checkAssertions = function()
 		assert(name and #name > 0, "Name is required")
-		assert(newProcessId and utils.isValidBase64Url(newProcessId), "Invalid Process-Id")
+		assert(utils.isValidAOAddress(newProcessId), "Process id must be a valid base64url.")
```
Comment (arielmelendez) [inline src/main.lua:1818]: This util function is too permissive (e.g. eth names).

**Review by arielmelendez — CHANGES_REQUESTED** [review]
See comment re: isvalidaoaddress

**Review by arielmelendez — APPROVED** [review]

**Pull request merged**